Add documentation for hot/cold splitting (#73029)
dotnet/runtime: docs/design/coreclr/jit/hot-cold-splitting.md

# Hot/Cold Code Splitting

This document describes the current state of hot/cold splitting in the JIT.

> Hot/Cold splitting is a PGO optimization where generated code is split into frequently-executed ("hot") and

Reviewer: Hot/cold splitting is not strictly Profile Guided Optimization. It is quite effective even without profile data.

> Hot/Cold splitting is an optimization where generated code is split into frequently-executed ("hot") and
> rarely-executed ("cold") parts and placed in separate memory regions. Increased hot code density better leverages

Reviewer 2: Nit -- keep an eye out for passive voice and try to avoid it where you can. Here it seems pretty easy:
```suggestion
Hot/Cold splitting is an optimization that splits code into frequently-executed ("hot") and
rarely-executed ("cold") parts and places the parts in separate memory regions. Increased hot code density better leverages
```

> spatial locality, which can improve application performance via fewer instruction cache misses, less OS paging, and

Reviewer 2: ```suggestion
spatial locality, improving application performance via fewer instruction cache misses, less OS paging, and
```

Reviewer 2: (you get the idea)

fewer TLB misses.

Hot/Cold splitting in the JIT was previously implemented for AOT-compiled NGEN images in .NET Framework. With hot/cold
splitting support in Crossgen2 [in progress](https://github.com/dotnet/runtimelab/tree/feature/hot-cold-splitting)
(and no existing support for splitting dynamically-generated code), the feature has gone untested in the JIT since
retiring .NET Framework -- as such, regressions have surely been introduced. Furthermore, this existing implementation
has several major limitations; functions with certain features, like exception handling or switch tables, cannot be
split yet. Finally, with ARM64/x64 performance parity being a high priority for CoreCLR, it is key hot/cold splitting

> is implemented for ARM64 code generation, too.

Reviewer 2: I think somebody could get confused by the mixed use of tense here and not be entirely sure what the "current state" actually is. 

Consider sectioning this part off as `## Background` and describing just the before state in past tense.

The below sections describe various improvements made to the JIT's hot/cold splitting support.

## Testing the JIT Without Runtime Support

> Without runtime support for hot/cold splitting in .NET Core yet, testing the JIT's existing hot/cold splitting support

Reviewer: ```suggestion
Without runtime support for hot/cold splitting in .NET as of summer 2022, testing the JIT's existing hot/cold splitting support
```
Nit: We have retired `.NET Core`  as the name a few years ago. We are calling the product just `.NET` now.

applicability.

Because EH funclets immediately succeed the main function body, the JIT can easily split such functions without
breaking existing invariants:

* If the JIT finds a split point in the main body, it splits there as usual. The latter part of the main body,
along with all of the function's EH funclets, are moved to the cold section.
* If the JIT does not find a split point in the main body, and none of the funclets are run frequently, it splits
at the beginning of the funclet section. The main body is left hot, and all EH funclets are moved to the cold section.
* Else, nothing is split.

This approach may not be the most performant implementation: Splitting funclets individually could yield better
spatial locality. However, this would require re-arranging the order of funclets (currently, no specific order is
imposed), and significantly altering how unwind info is generated, thus breaking many invariants in the host. This
approach enables splitting in many more scenarios without breaking existing invariants or introducing
architecture-specific workarounds. However, if the JIT supports splitting functions multiple times in the future, this
may be worth revisiting.

> In the absence of PGO data, the JIT assumes exceptions occur rarely to justify moving handlers to the cold section.

Reviewer 2: ```suggestion
In the absence of PGO data, the JIT assumes exceptions occur rarely; this justifies moving handlers to the cold section.
```
